windows.js

```diff
@@ -50,15 +54,70 @@ export async function install(platform, engine, version) {
 
   let toolchainPaths = (version === 'mswin') ? await setupMSWin() : await setupMingw(version)
 
-  common.setupPath([`${rubyPrefix}\\bin`, ...toolchainPaths])
-
   if (!inToolCache) {
     await downloadAndExtract(engine, version, url, base, rubyPrefix);
   }
 
+  const winMSYS2Type = common.setupPath([`${rubyPrefix}\\bin`, ...toolchainPaths])
+
+  // install msys2 tools for all Ruby versions, only install mingw or ucrt for Rubies >= 2.4
+
+  if (!['windows-2019', 'windows-2016'].includes(virtualEnv)) {
+    await installMSYS2Tools()
+  }
+
+  if ((( winMSYS2Type === 'ucrt64') || !['windows-2019', 'windows-2016'].includes(virtualEnv)) &&
```

Comment:
Extract `!['windows-2019', 'windows-2016'].includes(virtualEnv)` in a variable, for instance `const hasMSYSPreInstalled = ....` (usage here would be `!hasMSYSPreInstalled`)

Reply:
see above

Reply:
see above